fullnode/fullnode_cache.go

```diff
@@ -0,0 +1,228 @@
+	//   - TODO_TECHDEBT(@commoddity): Revisit based on real-world usage; consider making configurable
+	cacheCapacity = 100_000
+
+	// numShards:
+	//   - Number of independent cache shards for concurrency
+	//   - Reduces lock contention, improves parallelism
+	//   - 10 is a good balance for most workloads
+	numShards = 10
+
+	// evictionPercentage:
+	//   - % of LRU entries evicted per shard when full
+	//   - 10% = incremental cleanup, avoids memory spikes
+	//   - SturdyC also evicts expired entries in background
+	evictionPercentage = 10
+
+	// TODO_TECHDEBT(@commoddity): See Issue #291 for improvements to refresh logic
+	// minEarlyRefreshPercentage:
+	//   - Earliest point (as % of TTL) to start background refresh
+	//   - 0.75 = 75% of TTL (e.g. 22.5s for 30s TTL)
+	minEarlyRefreshPercentage = 0.75
+
+	// maxEarlyRefreshPercentage:
+	//   - Latest point (as % of TTL) to start background refresh
+	//   - 0.9 = 90% of TTL (e.g. 27s for 30s TTL)
+	//   - Ensures refresh always completes before expiry
+	maxEarlyRefreshPercentage = 0.9
+)
+
+// pubKeyCacheTTL: No TTL for the account public key cache since account data never changes.
+//
+// time.Duration(math.MaxInt64) equals ~292 years, which is effectively infinite.
+const pubKeyCacheTTL = time.Duration(math.MaxInt64)
+
+// getCacheDelays returns the min/max delays for SturdyC's Early Refresh strategy.
+// - Proactively refreshes cache before expiry (prevents misses/latency spikes)
+// - Refresh window: 75-90% of TTL (e.g. 22.5-27s for 30s TTL)
+// - Spreads requests to avoid thundering herd
+// See: https://github.com/viccon/sturdyc?tab=readme-ov-file#early-refreshes
+func getCacheDelays(ttl time.Duration) (min, max time.Duration) {
+	minFloat := float64(ttl) * minEarlyRefreshPercentage
+	maxFloat := float64(ttl) * maxEarlyRefreshPercentage
+
+	// Round to the nearest second
+	min = time.Duration(minFloat/float64(time.Second)+0.5) * time.Second
+	max = time.Duration(maxFloat/float64(time.Second)+0.5) * time.Second
+	return
+}
+
+// Prefix for cache keys to avoid collisions with other keys.
+const (
+	sessionCacheKeyPrefix       = "session"
+	accountPubKeyCacheKeyPrefix = "pubkey"
+)
+
+// cachingFullNode wraps a LazyFullNode with SturdyC-based caching.
+// - Early refresh: background updates before expiry (prevents thundering herd/latency spikes)
+// - Example: 30s TTL, refresh at 22.5–27s (75–90%)
+// - Benefits: zero-latency reads, graceful degradation, auto load balancing
+// Docs: https://github.com/viccon/sturdyc
+type cachingFullNode struct {
+	logger polylog.Logger
+
+	// Underlying node for protocol data fetches
+	lazyFullNode *LazyFullNode
+
+	// Session cache (5 min on Beta TestNet, see #275)
+	// TODO_MAINNET_MIGRATION(@Olshansk): Revisit after mainnet
+	sessionCache *sturdyc.Client[sessiontypes.Session]
+
+	// The account public key cache; used to cache account public keys indefinitely.
+	// It has an infinite TTL and is populated only once on startup.
```

Comment:
We don't know what will happen during runtime.

Inifnite TTL -> LGTM.
If there's a cache miss -> we should update JIT (just in time).

Reply:
That's currently how it works. SturdyC's `GetOrFetch` method handles remote fetching in the case of a cache miss.